nep-17.mediawiki

```diff
@@ -0,0 +1,109 @@
+
+==Specification==
+
+In the method definitions below, we provide both the definitions of the functions as they are defined in the contract as well as the invoke parameters.
+
+===Methods===
+
+====symbol====
+
+<pre>
+public static string symbol()
+</pre>
+
+Returns a short string symbol of the token managed in this contract. e.g. <code>"MYT"</code>. This symbol SHOULD be short (3-8 characters is recommended), with no whitespace characters or new-lines and SHOULD be limited to the uppercase latin alphabet (i.e. the 26 letters used in English).
+
+This method MUST always return the same value every time it is invoked.
+
+====decimals====
+
+<pre>
+public static byte decimals()
+</pre>
+
+Returns the number of decimals used by the token - e.g. <code>8</code>, means to divide the token amount by <code>100,000,000</code> to get its user representation.
+
+This method MUST always return the same value every time it is invoked.
+
+====totalSupply====
+
+<pre>
+public static BigInteger totalSupply()
+</pre>
+
+Returns the total token supply deployed in the system.
+
+====balanceOf====
+
+<pre>
+public static BigInteger balanceOf(byte[] account)
+</pre>
+
+Returns the token balance of the <code>account</code>.
+
+The parameter <code>account</code> SHOULD be a 20-byte address. If not, this method SHOULD <code>throw</code> an exception.
+
+If the <code>account</code> is an unused address, this method MUST return <code>0</code>.
+
+====transfer====
+
+<pre>
+public static bool transfer(byte[] from, byte[] to, BigInteger amount)
+</pre>
+
+Transfers an <code>amount</code> of tokens from the <code>from</code> account to the <code>to</code> account.
+
+The parameters <code>from</code> and <code>to</code> SHOULD be 20-byte addresses. If not, this method SHOULD <code>throw</code> an exception.
+
+The parameter <code>amount</code> MUST be greater than or equal to <code>0</code>. If not, this method SHOULD <code>throw</code> an exception.
+
+The function MUST return <code>false</code> if the <code>from</code> account balance does not have enough tokens to spend.
+
+If the method succeeds, it MUST fire the <code>transfer</code> event, and MUST return <code>true</code>, even if the <code>amount</code> is <code>0</code>, or <code>from</code> and <code>to</code> are the same address.
+
+The function SHOULD check whether the <code>from</code> address equals the caller contract hash. If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL <code>Neo.Runtime.CheckWitness</code> to verify the transfer.
+
+If the <code>to</code> address is a deployed contract, the function SHOULD check the <code>payable</code> flag of this contract to decide whether it should transfer the tokens to this contract.
+
+If the transfer is not processed, the function SHOULD return <code>false</code>.
```

Comment:
There are some optional behaviours (SHOULDs) where I wonder if they should be MUSTs instead
1. `from` and `to` SHOULD be 20-byte addresses and **SHOULD** throw an exception if not. I think the latter should be a MUST. Right now, if you pass a stack item with less then 20-bytes the ApplicationEngine will throw an exception because it can't convert the data to a `UInt160`. 
2. We say "if `transfer` succeeds it **MUST** return `true`", but at the same time we say "if transfer is not processed it **SHOULD** return `false`". Under the assumption that "not processed == transfer failed", then I think the latter should say "**MUST** return `false`"

Reply:
Fixed.